[DAG] Lower frem of power-2 using div/trunc/mul+sub

davemgreen opened the pull request:
If we are lowering a frem and the divisor is known to be an integer power-2, we
can use the formula 'frem = x - trunc(x / d) * d'. This avoids the more
expensive call to fmod. The results are identical as fmod so long as d is a
power-2 (so the mul does not round incorrectly), and the sign of the return is
either always positive or not important for zeroes (nsz).

Unfortunately Alive2 does not handle this well at the moment. I was using
exhaustive checking to test this, hopefully I didn't make a mistake in it
(https://gist.github.com/davemgreen/6078015f30d3bacd1e9572f8db5d4b64).

I found this in cpythons implementation of float_pow. I currently added it as a
DAG combine for frem with power-2 fp constants.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -17261,6 +17262,25 @@ SDValue DAGCombiner::visitFREM(SDNode *N) {
+        TLI.isOperationLegalOrCustom(ISD::FMUL, VT) &&
+        TLI.isOperationLegalOrCustom(ISD::FDIV, VT) &&
+        TLI.isOperationLegalOrCustom(ISD::FTRUNC, VT)) {
+      SDValue Div = DAG.getNode(ISD::FDIV, SDLoc(N), VT, N0, N1);
+      SDValue Rnd = DAG.getNode(ISD::FTRUNC, SDLoc(N), VT, Div);
+      SDValue Mul = DAG.getNode(ISD::FMUL, SDLoc(N), VT, Rnd, N1);
+      return DAG.getNode(ISD::FSUB, SDLoc(N), VT, N0, Mul);
```
Construct the SDLoc once? Also should probably be preserving all flags

arsenm commented on llvm/test/CodeGen/AArch64/frem-power2.ll:
```diff
@@ -0,0 +1,383 @@
+  %a = tail call <4 x float> @llvm.fabs.v4f32(<4 x float> %x)
+  %fmod = frem <4 x float> %a, <float 1152921504606846976.0, float 1152921504606846976.0, float 1152921504606846976.0, float 1152921504606846976.0>
+  ret <4 x float> %fmod
+}
```
Test the sitofp / uitofp cases

davemgreen replied:
I've added this and a few more, and tried adding some basic versions of isKnownNonNegativeFP and isKnownToBeAPowerOfTwoFP to handle those parts.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -17261,6 +17262,25 @@ SDValue DAGCombiner::visitFREM(SDNode *N) {
+  if (ConstantFPSDNode *C1 = isConstOrConstSplatFP(N1, true)) {
+    bool IsExact;
+    APSInt C1I(64, 0);
+    if (C1->getValueAPF().isInteger() && !C1->getValueAPF().isNegative() &&
+        C1->getValueAPF().convertToInteger(C1I, APFloat::rmTowardZero,
+                                           &IsExact) == APFloat::opOK &&
```
This should really be a separate isKnownInteger predicate, and also handle the cast-from-int case (also the various rounding intrinsics). I've been meaning to move  https://github.com/llvm/llvm-project/blob/30367cb5982dfdab2655401f020711311e7d78b9/llvm/lib/Target/AMDGPU/AMDGPULibCalls.cpp#L578

I've been meaning to move this to ValueTracking, but we need the equivalent DAG version

jayfoad commented:
This expansion should also work for negative power-of-two divisors like -4.0.

This expansion should also work for fractional power-of-two divisors like +/ 0.25, unless there is some concern about the division overflowing?

jayfoad commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -17261,6 +17262,25 @@ SDValue DAGCombiner::visitFREM(SDNode *N) {
     return NewSel;
 
+  if (ConstantFPSDNode *C1 = isConstOrConstSplatFP(N1, true)) {
+    bool IsExact;
+    APSInt C1I(64, 0);
+    if (C1->getValueAPF().isInteger() && !C1->getValueAPF().isNegative() &&
+        C1->getValueAPF().convertToInteger(C1I, APFloat::rmTowardZero,
+                                           &IsExact) == APFloat::opOK &&
+        IsExact && isPowerOf2_64(C1I.getSExtValue()) &&
```
```suggestion
    if (C1->getValueAPF().getExactLog2() >= 0 &&
```

davemgreen reviewed (commented):
> This expansion should also work for negative power-of-two divisors like -4.0.
> 
> This expansion should also work for fractional power-of-two divisors like +/ 0.25, unless there is some concern about the division overflowing?

Thanks for the suggestions. Negative power-2 sounds good -- I'll add those in. Values smaller than 1 unfortunately make it round to infinity in cases where the frem would be valid.

davemgreen committed:
[DAG] Lower frem of power-2 using div/trunk/mul+sub.

If we are lowering a frem and the divisor is known to me an integer power-2, we
can use the formula 'frem = x - trunc(x / d) * d'. This avoids the more
expensive call to fmod. The results are identical as fmod so long as d is a
power-2 (so the mul does not round incorrectly), and the sign of the return is
either always positive or signed-zero not important (nsz).

Unfortunately Alive2 does not handle this well at the moment. I was using
exhaustive checking to test this, hopefully I didn't make a mistake in it
(https://gist.github.com/davemgreen/6078015f30d3bacd1e9572f8db5d4b64).

I found this in cpythons implementation of float_pow. I currently added it as a
DAG combine for frem with power-2 fp constants, with some extra utility
functions for checking if a floating-point value is known non-negative or a
integer power-2.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -17264,17 +17264,32 @@ SDValue DAGCombiner::visitFREM(SDNode *N) {
+  // power of 2.
+  if (DAG.isKnownToBeAPowerOfTwoFP(N1) &&
+      (Flags.hasNoSignedZeros() || DAG.isKnownNonNegativeFP(N0)) &&
+      !TLI.isOperationLegal(ISD::FREM, VT) &&
```
Check the legality stuff first

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -17264,17 +17264,32 @@ SDValue DAGCombiner::visitFREM(SDNode *N) {
+    SDValue Div = DAG.getNode(ISD::FDIV, DL, VT, N0, N1);
+    SDValue Rnd = DAG.getNode(ISD::FTRUNC, DL, VT, Div);
+    SDValue Mul = DAG.getNode(ISD::FMUL, DL, VT, Rnd, N1);
+    return DAG.getNode(ISD::FSUB, DL, VT, N0, Mul);
```
Preserve the flags

arsenm replied:
One thing is you should be able to emit this fmul+fsub as an FMA + fneg. I was thinking you could introduce contract, but that's potentially wrong depending on downstream uses. 

You probably need to directly check isFMAFasterThanFMulAndFAdd (and maybe if isFNegFree) and emit that instead of fmul+fsub

davemgreen committed:
Do the legality checks first

arsenm commented on llvm/include/llvm/CodeGen/SelectionDAG.h:
```diff
@@ -2111,6 +2115,10 @@ class SelectionDAG {
 
+  /// Test whether the given float value is known to not be negative. 0.0 is
+  /// considered non-negative, -0.0 is considered negative.
+  bool isKnownNonNegativeFP(SDValue Op) const;
```
Someday we'll have to reinvent computeKnownFPClass in the DAG... (maybe we could do this in CGP to avoid that)

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5555,6 +5565,13 @@ bool SelectionDAG::isKnownNeverZero(SDValue Op, unsigned Depth) const {
 
+bool SelectionDAG::isKnownNonNegativeFP(SDValue Op) const {
+  if (ConstantFPSDNode *C1 = isConstOrConstSplatFP(Op, true))
+    return !C1->isNegative();
```
Maybe need to guard against nans

jayfoad replied:
Since nans are allowed I would suggest being explicit and renaming to `isKnownPositiveOrNaNFP`

arsenm reviewed (commented):
Basically LGTM, except should try to emit FMA if possible

davemgreen committed:
Use FMA

jayfoad commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -17289,8 +17289,13 @@ SDValue DAGCombiner::visitFREM(SDNode *N) {
+    if (TLI.isFMAFasterThanFMulAndFAdd(DAG.getMachineFunction(), VT)) {
+      return DAG.getNode(ISD::FMA, DL, VT, DAG.getNode(ISD::FNEG, DL, VT, Rnd),
+                         N1, N0);
+    } else {
```
No `else` after `return`

jayfoad commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5567,7 +5567,7 @@ bool SelectionDAG::isKnownNeverZero(SDValue Op, unsigned Depth) const {
 bool SelectionDAG::isKnownNonNegativeFP(SDValue Op) const {
   if (ConstantFPSDNode *C1 = isConstOrConstSplatFP(Op, true))
-    return !C1->isNegative();
+    return !C1->isNegative() && !C1->isNaN();
```
I'm not sure this is appropriate since the result of the FABS case below could still be NaN. What is the exact spec of this function?

davemgreen committed:
No else after return

arsenm commented on llvm/include/llvm/CodeGen/SelectionDAG.h:
```diff
@@ -2111,6 +2115,10 @@ class SelectionDAG {
 
+  /// Test whether the given float value is known to be positive. +0.0, +inf and
+  /// +nan are considered positive, -0.0, -inf and -nan are not.
+  bool isKnownPositiveOrNaNFP(SDValue Op) const;
```
Best to follow the ValueTracking naming and call this cannotBeOrderedLessThanZero

jayfoad replied:
Right, you can only call it that if you're happy with the semantics change too, which is that it would return true for -0.0.

arsenm replied:
> bitwise negative

Not quite, sign bit of nan is meaningless. How about cannotBeOrderedNegativeFP?

davemgreen replied:
I see, The IEEE 754 standard specifies that the sign bit of a NaN has no significance. The Arm architecture is obviously more opinionated about what the correct answer is, and I have been testing bitwise.

That probably helps with copysign too. I can change it to the new name.

davemgreen committed:
cannotBeOrderedNegativeFP

efriedma-quic commented on llvm/test/CodeGen/AArch64/frem-power2.ll:
```diff
@@ -13,31 +13,57 @@ entry:
+; CHECK-SD-LABEL: frem2_nsz:
+; CHECK-SD:       // %bb.0: // %entry
+; CHECK-SD-NEXT:    fmov s1, #2.00000000
+; CHECK-SD-NEXT:    fdiv s2, s0, s1
```
Is there some reason to prefer this sequence over `fmov s1, #0.5; fmul s2, s0, s1`?

davemgreen replied:
Sounds like a nice idea. It looks like there is an instcombiner for transforming that, but no equivalent in the backend at the moment. I'll try and take a look.
